src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,790 @@
+  }
+};
+
+// Find the minimum value that is not less than lo and satisfies bits.
+// Here, we view a number in binary as a bit string. As a result,  the first
+// bit refers to the highest bit (the MSB), the last bit refers to the lowest
+// bit (the LSB), a bit comes before (being higher than) another if it is more
+// significant, and a bit comes after (being lower than) another if it is less
+// significant.
+template <class U>
+static U adjust_lo(U lo, const KnownBits<U>& bits) {
+  constexpr size_t W = sizeof(U) * 8;
+  // Violation of lo with respects to bits
+  // E.g: lo    = 1100
+  //      zeros = 0100
+  //      ones  = 1001
+  // zero_violation = 0100, i.e the second bit should be zero, but it is 1 in
+  // lo. Similarly, one_violation = 0001, i.e the last bit should be one, but
+  // it is 0 in lo. These make lo not satisfy the bit constraints, which
+  // results in us having to find the smallest value that satisfies bits
+  U zero_violation = lo & bits._zeros;
+  U one_violation = ~lo & bits._ones;
+  if (zero_violation == one_violation) {
+    // This means lo does not violate bits, it is the result
+    assert(zero_violation == 0, "");
+    return lo;
+  }
+
+  // The principle here is that, consider the first bit in result that is
+  // different from the corresponding bit in lo, since result is larger than lo
+  // the bit must be 0 in lo and 1 in result. As result should be the smallest
+  // value, this bit should be the last one possible.
+  // E.g:      1 2 3 4 5 6
+  //      lo = 1 0 0 1 1 0
+  //       x = 1 0 1 0 1 0
+  //       y = 0 1 1 1 1 1
+  // x would be larger than lo since the first different bit is the 3rd one,
+  // while y is smaller than lo because the first different bit is the 1st bit.
+  // Next consider:
+  //      x1 = 1 0 1 0 1 0
+  //      x2 = 1 0 0 1 1 1
+  // Both x1 and x2 are larger than lo, but x1 > x2 since its first different
+  // bit from lo is the 3rd one, while with x2 it is the 7th one. As a result,
+  // if both x1 and x2 satisfy bits, x2 would be closer to our true result.
+
+  // The algorithm depends on whether the first violation violates zeros or
+  // ones, if it violates zeros, we have the bit being 1 in zero_violation and
+  // 0 in one_violation. Since all higher bits are 0 in zero_violation and
+  // one_violation, we have zero_violation > one_violation. Similarly, if the
+  // first violation violates ones, we have zero_violation < one_violation.
+  if (zero_violation < one_violation) {
+    // This means that the first bit that does not satisfy the bit requirement
+    // is a 0 that should be a 1, this may be the first different bit we want
+    // to find. The smallest value higher than lo with this bit being 1 would
+    // have all lower bits being 0. This value satisfies zeros, because all
+    // bits before the first violation have already satisfied zeros, and all
+    // bits after the first violation are 0. To satisfy 1, simply | this value
```

Comment:
```suggestion
    // bits after the first violation are 0. To satisfy ones, simply OR this value
```

Reply:
Ok, this I can understand much better!

Reply:
Maybe you can even add a line like below, to visually describe the result of the math below:
`result = lo[>i], 1, ones[<i]`